api/v1beta2/alert_types.go

```diff
@@ -50,10 +50,12 @@ type AlertSpec struct {
 	// +optional
 	InclusionList []string `json:"inclusionList,omitempty"`
 
-	// EventMetadata is an optional field for adding metadata to events emitted by the
-	// controller. Metadata fields added by the controller have priority over the fields
-	// added here, and the fields added here have priority over fields originally present
-	// in the event.
+	// EventMetadata is an optional field for adding metadata to events dispatched
+	// by the controller. The fields in this map have priority over the metadata
+	// originally present in the event, as issued by the emitter. The controller may
+	// still add specific fields to the final event that will be dispatched (for
+	// example, the "summary" field when present in the Alert spec), and these will
+	// have priority over everything else.
```

Comment:
The user has control over what's passed (statically) in this Alert spec field and the behavior of the field is now well documented after this pull request. If they choose to override a field that was cleaned up or that interfered with the rate limiting calculation, it's gonna be an informed decision. This field doesn't interfere with the rate limiting behavior, which is backwards compatible, which is good. I don't see a big problem here.

However, I do agree that we could maybe create a new field in the event payload here called `alert_metadata` and use it instead of `metadata`. The intention here is to deliver static metadata associated with the `Alert` object, not with the event.

Reply:
Maybe `eventMetadata` was a bad name choice :thinking:

Reply:
What if we delivered the labels or annotations of the `Alert` object as `alert_metadata` instead?

Reply:
I don't think `spec.eventMetadata` should be overriding the original event's metadata. It should only be adding more context. We could check if the key (from `spec.metadata`) is already in the event's metadata. If it is, we skip it and log a warning.

Reply:
> it could also be the intended behavior depending on the use case

Can you give some examples of such use cases? I am not sure why someone might want to override the metadata of an event intentionally. Events are updates from the flux controllers and overriding them could eventually lead to some confusion. They get sent to notification channels and are seen by other people (asides from the creator of the alert) who might not be aware that some parts have been overridden

Reply:
One may want to override user-defined metadata statically configured on the CR objects originating the events. That's not possible now but should be

Reply:
Yeah sorry, I haven't been super specific with my uses cases so far 🙁 Let me do that now.

First of all, the overriding behavior that I mentioned in my previous comment is not in my use cases, I just mentioned it because I thought it made sense in sight of also being able to specify static metadata on the originating CR object. I want to specify static metadata both in the originating CR object and on the `Alert` object, but I don't have a use case for overriding fields of the former with fields of the latter, at least not yet.

The fields I want to have as metadata are:

* `cluster`/`env`, e.g. prod/stag
* `component_dir`: one of `apps`, `infrastructure/configs` or `infrastructure/controllers`
* `component_name`: the name/path of an app or cluster addon/config

I want `cluster` to be specified on the `Alert` object such that I can have a single `Alert` object per cluster, and the other two on the `HelmRelease` objects.

Reply:
Okay, I will file PRs for fixing those things!

Yep! @stefanprodan mentioned that we need a new Event API version for supporting static metadata on the originating objects, and that it will only happen on a quarter further ahead in the year